cpp/src/arrow/vendored/fast_float/parse_number.h

```diff
@@ -1,5 +1,6 @@
 #ifndef FASTFLOAT_PARSE_NUMBER_H
 #define FASTFLOAT_PARSE_NUMBER_H
+#define FASTFLOAT_ALLOWS_LEADING_PLUS 1
```

Comment:
Could you define this in `cpp/src/arrow/util/value_parsing.cc` instead of here?
If we define this here, this change will be discarded when we update fast_float next time.
And could you also add a Gandiva test to check `+Infinity` is accepted or not?